staking/parse.go

```diff
@@ -3,13 +3,215 @@
+
+	var spk asn1.BitString
+	if !input.ReadASN1BitString(&spk) {
+		return nil, ErrMalformedSubjectPublicKey
+	}
+	publicKey, signatureAlgorithm, err := parsePublicKey(&publicKeyInfo{
+		Algorithm: pkAI,
+		PublicKey: spk,
+	})
+	return &Certificate{
+		Raw:                bytes,
+		SignatureAlgorithm: signatureAlgorithm,
+		PublicKey:          publicKey,
+	}, err
+}
+
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/parser.go#L149-L165
+func parseAI(der cryptobyte.String) (pkix.AlgorithmIdentifier, error) {
+	ai := pkix.AlgorithmIdentifier{}
+	if !der.ReadASN1ObjectIdentifier(&ai.Algorithm) {
+		return ai, ErrMalformedOID
+	}
+	if der.Empty() {
+		return ai, nil
+	}
+
+	var (
+		params cryptobyte.String
+		tag    cryptobyte_asn1.Tag
+	)
+	if !der.ReadAnyASN1Element(&params, &tag) {
+		return ai, ErrMalformedParameters
+	}
+	ai.Parameters.Tag = int(tag)
+	ai.Parameters.FullBytes = params
+	return ai, nil
+}
+
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/x509.go#L172-L176
+type publicKeyInfo struct {
+	Algorithm pkix.AlgorithmIdentifier
+	PublicKey asn1.BitString
+}
+
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/parser.go#L215-L306
+func parsePublicKey(keyData *publicKeyInfo) (any, x509.SignatureAlgorithm, error) {
+	oid := keyData.Algorithm.Algorithm
+	params := keyData.Algorithm.Parameters
+	der := cryptobyte.String(keyData.PublicKey.RightAlign())
+	switch {
+	case oid.Equal(oidPublicKeyRSA):
+		// RSA public keys must have a NULL in the parameters.
+		// See RFC 3279, Section 2.3.1.
+		if !bytes.Equal(params.FullBytes, asn1.NullBytes) {
+			return nil, 0, ErrRSAKeyMissingNULLParameters
+		}
+
+		pub := &rsa.PublicKey{N: new(big.Int)}
+		if !der.ReadASN1(&der, cryptobyte_asn1.SEQUENCE) {
+			return nil, 0, ErrInvalidRSAPublicKey
+		}
+		if !der.ReadASN1Integer(pub.N) {
+			return nil, 0, ErrInvalidRSAModulus
+		}
+		if !der.ReadASN1Integer(&pub.E) {
+			return nil, 0, ErrInvalidRSAPublicExponent
+		}
+
+		if pub.N.Sign() <= 0 {
+			return nil, 0, ErrRSAModulusNotPositive
+		}
+		// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/tls/handshake_client.go#L874-L877
+		if bitLen := pub.N.BitLen(); bitLen > MaxRSAKeyBitLen {
+			return nil, 0, fmt.Errorf("%w: bitLen=%d > maxBitLen=%d", ErrInvalidRSAPublicKey, bitLen, MaxRSAKeyBitLen)
+		}
+		if pub.E <= 0 {
+			return nil, 0, ErrRSAPublicExponentNotPositive
+		}
+		return pub, x509.SHA256WithRSA, nil
+	case oid.Equal(oidPublicKeyECDSA):
+		paramsDer := cryptobyte.String(params.FullBytes)
+		var namedCurveOID asn1.ObjectIdentifier
+		if !paramsDer.ReadASN1ObjectIdentifier(&namedCurveOID) {
+			return nil, 0, ErrInvalidECDSAParameters
+		}
+		// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/x509.go#L491-L503
+		if !namedCurveOID.Equal(oidNamedCurveP256) {
+			return nil, 0, ErrUnsupportedEllipticCurve
+		}
+		namedCurve := elliptic.P256()
+		x, y := elliptic.Unmarshal(namedCurve, der)
```

Comment:
TODO: Should we use a different parsing function? This aligns with the x509 package, but is deprecated.

Reply:
The canonical entry point is x509.ParsePKIXPublicKey but you are correctly reimplementing that. I think this might be a rare enough case in which elliptic.Unmarshal is appropriate. Its behavior is very well specified, anyway, so you don't need to worry about it changing.